### PR TITLE
Add featuregate pattern to scheduler framework handle

### PR DESCRIPTION
**Pull request by damemi**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This is an alternative to https://github.com/kubernetes/kubernetes/pull/98545 and https://github.com/kubernetes/kubernetes/pull/98583, which both attempt to set a design for removing a dependency on feature gate logic from individual plugins. This PR does that by adding options directly to the framework handle, to be called at initialization of the framework.

**Which issue(s) this PR fixes**:

Ref #89930

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Scheduler framework handle interface now includes FeatureGates() function to provide feature gate information to plugins
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```
/sig scheduling

### CODE REVIEW

File: pkg/scheduler/framework/interface.go [hunk 1]
```diff
@@ -566,6 +567,9 @@ type Handle interface {
 
 	// TODO: unroll the wrapped interfaces to Handle.
 	PreemptHandle() PreemptHandle
+
+	// FeatureGates returns if a featuregate is registered and enabled for this handle
+	FeatureGates(feature featuregate.Feature) bool
```
Comment (alculquicondor) [inline pkg/scheduler/framework/interface.go:572]: `FeatureGateEnabled` or just `FeatureEnabled`

File: pkg/scheduler/framework/plugins/noderesources/fit_test.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ package noderesources
 import (
 	"context"
 	"fmt"
+	"k8s.io/kubernetes/pkg/scheduler/framework/runtime"
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/noderesources/fit_test.go:22]: move to 2nd import block

File: pkg/scheduler/framework/runtime/framework.go [hunk 3]
```diff
@@ -403,6 +423,14 @@ func updatePluginList(pluginList interface{}, pluginSet *config.PluginSet, plugi
 	return nil
 }
 
+// FeatureGates returns whether a feature gate is registered and enabled with the handle
+func (f *frameworkImpl) FeatureGates(feature featuregate.Feature) bool {
```
Comment (alculquicondor) [inline pkg/scheduler/framework/runtime/framework.go:427]: This package is already depending on the `utilfeature` package, so I don't see the need to have this map.

Just call `DefaultFeatureGate.Enabled` here and remove the `With` functions

File: pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go [hunk 4]
```diff
@@ -88,7 +88,7 @@ func balancedResourceScorer(requested, allocable resourceToValueMap, includeVolu
 		return 0
 	}
 
-	if includeVolumes && utilfeature.DefaultFeatureGate.Enabled(features.BalanceAttachedNodeVolumes) && allocatableVolumes > 0 {
+	if includeVolumes && allocatableVolumes > 0 {
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go:91]: do we not need to test the flag anymore?

**Review by ahg-g — COMMENTED** [review]
the constants are not in the component-base pkg, so I am not sure this whole approach will work if we don't duplicate the constants definitions, which I am not sure we should be doing.

File: pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go (hunk 4 above)
Comment (damemi) [reply pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go:91]: we don't, the calling function only sets `includeVolumes` if the flag is enabled. Forgot to carry that comment over https://github.com/kubernetes/kubernetes/pull/98583/files#diff-6fe34bf325186cb8568837e8c946367248be19d153f9e4ce1b05b1d3e66ecc10R109

**Comment by damemi** [comment]
> the constants are not in the component-base pkg, so I am not sure this whole approach will work if we don't duplicate the constants definitions, which I am not sure we should be doing.

I was thinking about this too. If we could abstract the constants away behind interface functions, it might work.

**Comment by damemi** [comment]
Closing this in favor of https://github.com/kubernetes/kubernetes/pull/98583#issuecomment-772693619
/close

**Pull request closed without merging**